canals/pipeline/connections.py

```diff
@@ -72,13 +72,59 @@ def _connections_status(from_node: str, to_node: str, from_sockets: List[OutputS
     """
     Lists the status of the sockets, for error messages.
     """
-    from_sockets_list = "\n".join(
-        [f" - {socket.name} ({[t.__name__ for t in socket.types]})" for socket in from_sockets]
-    )
-    to_sockets_list = "\n".join(
-        [
-            f" - {socket.name} ({[t.__name__ for t in socket.types]}, {'sent by '+socket.sender if socket.sender else 'available'})"
-            for socket in to_sockets
-        ]
-    )
+    from_sockets_entries = []
+    for from_socket in from_sockets:
+        socket_types = ", ".join([_get_socket_type_desc(t) for t in from_socket.types])
+        from_sockets_entries.append(f" - {from_socket.name} ({socket_types})")
+    from_sockets_list = "\n".join(from_sockets_entries)
+
+    to_sockets_entries = []
+    for to_socket in to_sockets:
+        socket_types = ", ".join([_get_socket_type_desc(t) for t in to_socket.types])
+        to_sockets_entries.append(
+            f" - {to_socket.name} ({socket_types}), {'sent by '+to_socket.sender if to_socket.sender else 'available'}"
+        )
+    to_sockets_list = "\n".join(to_sockets_entries)
+
     return f"'{from_node}':\n{from_sockets_list}\n'{to_node}':\n{to_sockets_list}"
+
+
+def _get_socket_type_desc(type_):
+    """
+    Assembles a readable representation of the type of a connection. Can handle primitive types, classes, and
+    arbitrarily nested structures of types from the typing module.
+    """
+    # get_args returns something only if this type has subtypes, in which case it needs to be printed differently.
+    args = get_args(type_)
+
+    if not args:
+        if isinstance(type_, type):
+            if type_.__name__.startswith("typing."):
+                return type_.__name__[len("typing.") :]
+            return type_.__name__
+
+        # Literals only accept instances, not classes, so we need to account for those.
+        if isinstance(type_, str):
+            return f"'{type_}'"  # Quote strings
+        if str(type_).startswith("typing."):
+            return str(type_)[len("typing.") :]
+        return str(type_)
+
+    # Python < 3.10 support
+    if hasattr(type_, "_name"):
+        type_name = type_._name  # pylint: disable=protected-access
+        # Support for Optionals and Unions in Python < 3.10
+        if not type_name:
+            if type(None) in args:
+                type_name = "Optional"
+                args = [a for a in args if a is not type(None)]
+            else:
+                if not any(isinstance(a, type) for a in args):
+                    type_name = "Literal"
+                else:
+                    type_name = "Union"
```

Comment:
Can we maybe use `typing_extensions` instead of this? 🤔

Reply:
Types come from the components, so unless we want to force users to use `typing_extensions` (I wouldn't) that's won't help I think